### PR TITLE
Allow data passed from editor to preview

### PR DESCRIPTION
**Pull request by thomasloven**
## Proposed change

Allow card editors to communicate state to the card previews.

U7KuLLqEqV

Implemented in `entities`, `-stack` and `grid` cards.

## Type of change

- [ ] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [X] New feature (thank you!)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Example configuration

```yaml

```

## Additional information

- This PR fixes or closes issue: fixes #
- This PR is related to issue or discussion:
- Link to documentation pull request:

## Checklist

- [X] The code change is tested and works locally.
- [X] There is no commented out code in this PR.
- [ ] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [ ] Documentation added/updated for [www.home-assistant.io][docs-repository]

[docs-repository]: https://github.com/home-assistant/home-assistant.io

### CODE REVIEW

File: src/panels/lovelace/cards/hui-stack-card.ts [hunk 1]
```diff
@@ -43,8 +44,16 @@ export abstract class HuiStackCard<T extends StackCardConfig = StackCardConfig>
...
+    this._cards = config.cards.map((card, idx) => {
       const element = this._createCardElement(card) as LovelaceCard;
+      if (this.editMode !== undefined) {
+        if (this.editMode?.selected === idx) {
+          element.classList.add("selected");
+          element.editMode = this.editMode.data;
+        } else {
+          element.editMode = true;
+        }
+      }
```
Comment (zsarnett) [inline src/panels/lovelace/cards/hui-stack-card.ts:56]:
```suggestion
    if (this.editMode?.selected === idx) {
      element.classList.add("selected");
      element.editMode = this.editMode.data;
    } else {
      element.editMode = true;
    }
```

File: src/panels/lovelace/types.ts [hunk 2]
```diff
@@ -38,7 +38,7 @@ export interface LovelaceBadge extends HTMLElement {
 export interface LovelaceCard extends HTMLElement {
   hass?: HomeAssistant;
   isPanel?: boolean;
-  editMode?: boolean;
+  editMode?: any;
```
Comment (zsarnett) [inline src/panels/lovelace/types.ts:41]: Let's go ahead and type this to 

```suggestion
  editMode?: boolean | EditMode;
```

```ts
interface EditMode {
	selectedRow: Number;
	selectedCard: Number;
}
```

Comment (thomasloven) [reply src/panels/lovelace/types.ts:41]: I want to limit it as little as possible. Custom cards should be able to use this too and populate with whatever data they want.

File: src/panels/lovelace/cards/hui-stack-card.ts (hunk 1 above)
[the remark this answers is not in the document]
Comment (thomasloven) [reply src/panels/lovelace/cards/hui-stack-card.ts:56]: I think we still want to set editMode to true for all the other cards in the stack, such that e.g. conditional cards show up.

**Comment by angelfeliz** [comment]
I'm new to this project and at the moment, I'm only watching how the PR, Issue and discussion work. Being that said, I have one question. Should there exist a related discussion or Issue for this PR?

**Comment by spacegaier** [comment]
> I'm new to this project and at the moment, I'm only watching how the PR, Issue and discussion work. Being that said, I have one question. Should there exist a related discussion or Issue for this PR?

That is not required. In this instance there was a discussion in the Discord #dev_frontend channel though.

File: src/panels/lovelace/types.ts (hunk 2 above)
Comment (zsarnett) [reply src/panels/lovelace/types.ts:41]:
```ts
interface EditMode {
    selectedRow: Number;
    selectedCard: Number;
    [key: string]: any;
}
```

Comment (zsarnett) [reply src/panels/lovelace/types.ts:41]: or we should create a part of the edit mode that can be any

```ts
interface EditMode {
    selectedRow: Number;
    selectedCard: Number;
    data: any; // But not data as the key
}

```

**Comment by thomasloven** [comment]
I guess I'll bump this. It's not forgotten.

**Comment by kinghat** [comment]
keeping fresh

**Comment by frenck** [comment]
We have discussed this, we do like it, and we think we should improve the nested editor to be more user-friendly when editing a card in a grid. But, think this logic should be in the editor and not in the specific cards.

Going to close this PR for now.

../Frenck

**Comment by thomasloven** [comment]
I agree that would be optimal.
I hope someone can come up with a way to make that work.

**Pull request closed without merging**